### ISSUE TITLE
Prediction Model Not Working -Using Own Inputs in a KAN y equation keeps churning out same number no matter what inputs are used

### ISSUE THREAD

**Issue by vabudar** [KindXiaoming/pykan#393]
Hi, firstly thank you all very much for all the work done on this so far. It is really cutting edge and a massive discovery so well done.

I'm trying to use the KAN method as predictor of the compressive strength of concrete. I have all the inputs and all the compressive strengths in a CSV however after I've trained the model is seems the y value keeps the same regardless of the inputs, they're all around 45.......

This is the code I've used, is anything glaring wrong?:

from kan import *
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sympy as sp

model = KAN(width=[5,10,1], grid=500, k=3, seed=0) #KAN model size, 5 inputs, 10 neurons in hidden layer, 1 output

file_path = r"C:\Users\GA\OneDrive - University of Strathclyde\Masters\Dissertation\Concrete Data Shtuff.csv" #imports Concrete Data File

# Function to read CSV and extract columns within a given range, converting items to numbers
def inputs(file_path,):
    # Read the CSV file
    df = pd.read_csv(file_path)
    
    # Extract columns 3-7 (index 2-6) within the specified range
    extracted_data = df.iloc[1:49, 2:7].values
    
    # Convert all items in the extracted array to numbers
    numeric_data = np.array(extracted_data).astype(float)
    
    return numeric_data

def outputs(file_path, start_row, end_row):
    # Read the CSV file
    df = pd.read_csv(file_path)
    
    # Extract column 8 (index 7) within the specified range
    training_outputs = df.iloc[start_row:end_row, 7].values.astype(float)
    
    return training_outputs

    
df = pd.read_csv(file_path)
S_number=int(input("Please insert which sample you'd like to test for: "))
WCM_ratio=df.iloc[S_number, 2]
C_Content=df.iloc[S_number, 3]
Sand_Content=df.iloc[S_number, 4]
CA_Content=df.iloc[S_number, 5]
Cement_Content=df.iloc[S_number, 6]
day28_Compressive_Strength=df.iloc[S_number,7]

training_inputs = inputs(file_path)
training_outputs = outputs(file_path, 1, 49)

x_train = training_inputs.astype(np.float32)
y_train = training_outputs.astype(np.float32)
x_test = training_inputs.astype(np.float32)
y_test = training_outputs.astype(np.float32)

dataset = {}
dataset['train_input'] = torch.from_numpy(x_train)
dataset['test_input'] = torch.from_numpy(x_test)

dataset['train_label'] = torch.from_numpy(y_train)
dataset['test_label'] = torch.from_numpy(y_test)

model(dataset['train_input'])
results = model.fit(dataset, opt="LBFGS", steps=50, lamb=0.01, lamb_entropy=10)
model.plot(beta=1000)

train_rmse = []
test_rmse = []

for i in range(5):
    train_rmse.append(results['train_loss'][-1].item())
    test_rmse.append(results['test_loss'][-1].item())

lib = ['x','x^2','x^3','x^4','exp','log','sqrt','tanh','sin','tan','abs']
model.auto_symbolic(lib=lib)
formula = model.symbolic_formula()[0][0]
print(formula)

# Dictionary of variable values
variable_values = {
    'x_1': WCM_ratio,
    'x_2': C_Content,
    'x_3': Sand_Content,
    'x_4': CA_Content,
    'x_5': Cement_Content
}

 #new model hopefully retrained

model(dataset['train_input'])
model = model.prune()
results = model.fit(dataset, opt="LBFGS", steps=50, lamb=0.01, lamb_entropy=10)
model.plot(beta=1000)

formula = model.symbolic_formula()[0][0]
print(formula)
formula= str(formula)
# Substitute variables in the formula
for var, val in variable_values.items():
    formula = formula.replace(var, str(val))

# Evaluate the formula
result = eval(formula)

print("The formula after substitution:", formula)
print(f"Predicted Compressive Strength for Sample {S_number} is:", result)
print(f"Actual Compressive Strength for sample {S_number} is: {day28_Compressive_Strength}")

print(train_rmse)
print(test_rmse)

[output] Predicted Compressive Strength for Sample 4 is: 45.7450918077127
Actual Compressive Strength for sample 4 is: 47.16

My regression and train_loss/test_loss is quite high:

| train_loss: 5.38e+00 | test_loss: 5.38e+00 | reg: 8.40e+01 | : 100%|█| 50/50

Might it have something to do with that? 

Many thanks in advance for your help

**Comment by vabudar** [comment]
Solution found! The training_outputs had the wrong transform this resolves it 

training_outputs = torch.tensor(extracted_data, dtype=torch.float32).unsqueeze(1)
